torch/utils/data/dataloader.py

```diff
@@ -41,20 +41,28 @@ def create_fetcher(kind, dataset, auto_collation, collate_fn, drop_last):
 class _InfiniteConstantSampler(Sampler):
     r"""Analogous to ``itertools.repeat(None, None)``.
     Used as sampler for :class:`~torch.utils.data.IterableDataset`.
+
+    Arguments:
+        data_source (Dataset): dataset to sample from
     """
 
-    def __init__(self):
-        super(_InfiniteConstantSampler, self).__init__(None)
+    def __init__(self, data_source):
+        self.data_source = data_source
 
     def __iter__(self):
         while True:
             yield None
 
     def __len__(self):
-        # This has to be a TypeError, otherwise, since this is used in
-        # `len(dataloader)`, `list(dataloader)` will fail.
-        # see NOTE [ Lack of Default `__len__` in Python Abstract Base Classes ]
-        raise TypeError('Cannot determine the DataLoader length of a IterableDataset')
+        # Note that this could be inaccurate when one naively does
+        # multi-processing data loading, since the samples will be duplicated.
+        # However, no real use case should be actually using that behavior, so
+        # it should count as a user error. We should generally trust dataset
+        # code to do the proper thing (e.g., configure each replica differently
+        # in `__iter__`), and give us the correct `__len__` if they choose to
+        # implement it (this will still throw if the dataset does not implement
+        # a `__len__`).
+        return len(self.data_source)
```

Comment:
Naively, I would have expected `__len__` and `__iter__` to agree here. But iter never returns. So how can there be a length? Am I misunderstanding the relationship between these magic methods?

Reply:
Ah sorry for the confusion. I didn't explain this clear enough. 

`len(data_loader)` directly returns the `len(self.index_sampler)`. Previously,  just throws when using an `IterableDataset` because an `_InfiniteConstantSampler` is used for the dataset and its `__len__` throws. 

However, the `IterableDataset` provided by users could already implement `__len__` itself. So what this patch does is that `_InfiniteConstantSampler.__len__` now grabs the `__len__` of the dataset (and throw if it is not implemented) so the data loader can have the provided `__len__`.

Reply:
OK, that makes sense, but it doesn't (as far as I can tell), explain what I was confused about in the first place, which is how `__iter__` relates to `__len__`?

Reply:
They are not. We trust users to give a correct `__len__` that matches what their `__iter__` does, similar to what we did for old map-style dataset, i.e., trusting the dataset `__len__` matches what their `__getitem__` supports.